src/misc/programs.cpp

```diff
@@ -39,43 +41,51 @@ Bitu call_program;
 
 /* This registers a file on the virtual drive and creates the correct structure for it*/
 
-static Bit8u exe_block[]={
-	0xbc,0x00,0x04,					//MOV SP,0x400 decrease stack size
-	0xbb,0x40,0x00,					//MOV BX,0x040 for memory resize
-	0xb4,0x4a,						//MOV AH,0x4A	Resize memory block
-	0xcd,0x21,						//INT 0x21
-//pos 12 is callback number
-	0xFE,0x38,0x00,0x00,			//CALLBack number
-	0xb8,0x00,0x4c,					//Mov ax,4c00
-	0xcd,0x21,						//INT 0x21
+constexpr uint8_t exe_block[] = {
+        0xbc, 0x00, 0x04,       // MOV SP,0x400  Decrease stack size
+        0xbb, 0x40, 0x00,       // MOV BX,0x040  For memory resize
+        0xb4, 0x4a,             // MOV AH,0x4A   Resize memory block
+        0xcd, 0x21,             // INT 0x21
+        0xFE, 0x38, 0x00, 0x00, // 12th position is the CALLBack number
+        0xb8, 0x00, 0x4c,       // Mov ax,4c00
+        0xcd, 0x21,             // INT 0x21
 };
 
-#define CB_POS 12
+// COM data constants
+constexpr int callback_pos = 12;
+constexpr int com_data_size = 32;
 
-static std::vector<PROGRAMS_Main*> internal_progs;
+// Persistent program containers
+using com_data_t = std::unique_ptr<uint8_t[]>;
+static std::vector<com_data_t> internal_progs_comdata;
+static std::vector<PROGRAMS_Main *> internal_progs;
 
-void PROGRAMS_MakeFile(char const * const name,PROGRAMS_Main * main) {
-	Bit8u * comdata=(Bit8u *)malloc(32); //MEM LEAK
+void PROGRAMS_MakeFile(const char *name, PROGRAMS_Main *main)
+{
+	com_data_t comdata = std::make_unique<uint8_t[]>(com_data_size);
+	assert(comdata);
 
-	if (comdata == nullptr) {
-		E_Exit("Could not allocate memory for com-file data.");
-	}
+	memcpy(comdata.get(), &exe_block, sizeof(exe_block));
+	comdata[callback_pos] = static_cast<uint8_t>(call_program & 0xff);
+	comdata[callback_pos + 1] = static_cast<uint8_t>((call_program >> 8) & 0xff);
 
-	memcpy(comdata,&exe_block,sizeof(exe_block));
-	comdata[CB_POS]=(Bit8u)(call_program&0xff);
-	comdata[CB_POS+1]=(Bit8u)((call_program>>8)&0xff);
+	// Save the current program's vector index in its COM data
+	assert(internal_progs.size() <= UINT8_MAX);
+	const auto index = static_cast<uint8_t>(internal_progs.size());
+	comdata[sizeof(exe_block)] = index;
```

Comment:
All functionally identical.
Note the `.get()` call on `comdata` gets the raw memory of the smart pointer, so we can use memcpy, etc on it.

Reply:
any desire to go further with the C++? For example:
```
constexpr std::array<uint8_t, 19> exe_block = {...}
```
then
```
	std::copy(exe_block.begin(), exe_block.end(), comdata.get());
```
and
```
	PhysPt reader=PhysMake(dos.psp(),256+exe_block.size());
```

etc.

Reply:
Thanks for the suggestion! Yes, let's go for it. 

There's only upside to `std::copy` - handles any data type or container (fails to compile when something's not right), and will match or exceed `memcpy()` in terms of performance.

fixups inbound.

Reply:
- switched `exec_block` to `std::array`
 - switched `comdata` from a `unique_ptr` allocation to a `vector`
    - I thought about using `std::array` for comdata too, but `vector` lets us "push_back" and is also `std::move`-able into the global container, where as `std::array` would need to be copied. 
    - We can initialize a std::vector from begin() to end(), so turns out std::copy wasn't needed

Overall safer and roughly the same in terms of readability :+1: 
Thanks, @kklobe!